README.md
# nom-both
Extension of [nom](https://github.com/Geal/nom) to provide special `both_` parsers.

[![Build Status](https://dev.azure.com/dalance/nom-both/_apis/build/status/dalance.nom-both?branchName=master)](https://dev.azure.com/dalance/nom-both/_build/latest?definitionId=1&branchName=master)
[![Crates.io](https://img.shields.io/crates/v/nom-both.svg)](https://crates.io/crates/nom-both)
[![Docs.rs](https://docs.rs/nom-both/badge.svg)](https://docs.rs/nom-both)

## Requirement

nom must be 5.0.0 or later.
nom-both can be applied to function-style parser only.

## Usage

```Cargo.toml
[dependencies]
nom-both = "0.1.1"
```

## Example

nom-both provide `both_opt` and `both_alt` parser.
`both_opt` means that the parser tries both argument parser and skip.
`both_alt` means that the parser tries both 1st argument parser and 2nd argument parser.

```rust
use nom::branch::*;
use nom::bytes::complete::*;
use nom::IResult;
use nom_both::both_parser;

#[both_parser]
pub fn both_opt_parser(s: &str) -> IResult<&str, Option<&str>> {
    let (s, _) = tag("1")(s)?;
    let (s, x) = both_opt(tag("2"))(s)?;
    let (s, _) = tag("2")(s)?;
    let (s, _) = tag("3")(s)?;
    Ok((s, x))
}

#[both_parser]
pub fn both_alt_parser(s: &str) -> IResult<&str, &str> {
    let (s, _) = tag("1")(s)?;
    let (s, x) = both_alt(tag("22"), tag("2"))(s)?;
    let (s, _) = tag("2")(s)?;
    let (s, _) = tag("3")(s)?;
    Ok((s, x))
}

#[test]
fn test() {
    let ret = both_opt_parser("123");
    assert_eq!("None", format!("{:?}", ret.unwrap().1));

    let ret = both_alt_parser("1223");
    assert_eq!("\"2\"", format!("{:?}", ret.unwrap().1));
}
```

`both_opt_parser` is expanded as below:

```rust
pub fn both_opt_parser(s: &str) -> IResult<&str, Option<&str>> {
    alt((
        { |s| {
            let (s, _) = tag("1")(s)?;
            let (s, x) = nom_both::some(tag("2"))(s)?;
            let (s, _) = tag("2")(s)?;
            let (s, _) = tag("3")(s)?;
            Ok((s, x))
        } },
        { |s| {
            let (s, _) = tag("1")(s)?;
            let (s, x) = nom_both::none(tag("2"))(s)?;
            let (s, _) = tag("2")(s)?;
            let (s, _) = tag("3")(s)?;
            Ok((s, x))
        } },
    ))(s)
}
```

`both_alt_parser` is expanded as below:

```rust
pub fn both_opt_parser(s: &str) -> IResult<&str, Option<&str>> {
    alt((
        { |s| {
            let (s, _) = tag("1")(s)?;
            let (s, x) = nom_both::alt_left(tag("22"), tag("2"))(s)?;
            let (s, _) = tag("2")(s)?;
            let (s, _) = tag("3")(s)?;
            Ok((s, x))
        } },
        { |s| {
            let (s, _) = tag("1")(s)?;
            let (s, x) = nom_both::alt_right(tag("22"), tag("2"))(s)?;
            let (s, _) = tag("2")(s)?;
            let (s, _) = tag("3")(s)?;
            Ok((s, x))
        } },
    ))(s)
}
```

## License

Licensed under either of

 * Apache License, Version 2.0, ([LICENSE-APACHE](LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license ([LICENSE-MIT](LICENSE-MIT) or http://opensource.org/licenses/MIT)

at your option.

### Contribution

Unless you explicitly state otherwise, any contribution intentionally
submitted for inclusion in the work by you, as defined in the Apache-2.0
license, shall be dual licensed as above, without any additional terms or
conditions.


CHANGELOG.md
# Change Log

## [Unreleased](https://github.com/dalance/nom-both/compare/v0.1.1...Unreleased) - ReleaseDate

## [v0.1.1](https://github.com/dalance/nom-both/compare/v0.1.0...v0.1.1) - 2019-07-29



nom-both/tests/test.rs
use nom::branch::*;
use nom::bytes::complete::*;
use nom::IResult;
use nom_both::both_parser;

#[both_parser]
pub fn both_opt_parser(s: &str) -> IResult<&str, Option<&str>> {
    let (s, _) = tag("1")(s)?;
    let (s, x) = both_opt(tag("2"))(s)?;
    let (s, _) = tag("2")(s)?;
    let (s, _) = tag("3")(s)?;
    Ok((s, x))
}

#[both_parser]
pub fn both_alt_parser(s: &str) -> IResult<&str, &str> {
    let (s, _) = tag("1")(s)?;
    let (s, x) = both_alt(tag("22"), tag("2"))(s)?;
    let (s, _) = tag("2")(s)?;
    let (s, _) = tag("3")(s)?;
    Ok((s, x))
}

#[test]
fn test() {
    let ret = both_opt_parser("123");
    assert_eq!("None", format!("{:?}", ret.unwrap().1));

    let ret = both_alt_parser("1223");
    assert_eq!("\"2\"", format!("{:?}", ret.unwrap().1));
}


nom-both/src/lib.rs
//! `nom-both` is an extension of [nom](https://docs.rs/nom) to provide special `both_` parser.
//!
//! ## Examples
//!
//! The following example show a quick example.
//!
//! ```
//! use nom::branch::*;
//! use nom::bytes::complete::*;
//! use nom::IResult;
//! use nom_both::both_parser;
//!
//! #[both_parser]
//! pub fn both_opt_parser(s: &str) -> IResult<&str, Option<&str>> {
//!     let (s, _) = tag("1")(s)?;
//!     let (s, x) = both_opt(tag("2"))(s)?;
//!     let (s, _) = tag("2")(s)?;
//!     let (s, _) = tag("3")(s)?;
//!     Ok((s, x))
//! }
//!
//! #[both_parser]
//! pub fn both_alt_parser(s: &str) -> IResult<&str, &str> {
//!     let (s, _) = tag("1")(s)?;
//!     let (s, x) = both_alt(tag("22"), tag("2"))(s)?;
//!     let (s, _) = tag("2")(s)?;
//!     let (s, _) = tag("3")(s)?;
//!     Ok((s, x))
//! }
//!
//! fn main() {
//!     let ret = both_opt_parser("123");
//!     assert_eq!("None", format!("{:?}", ret.unwrap().1));
//!
//!     let ret = both_alt_parser("1223");
//!     assert_eq!("\"2\"", format!("{:?}", ret.unwrap().1));
//! }
//! ```

use nom::IResult;
pub use nom_both_macros::both_parser;

pub fn some<'a, I, O, F>(f: F) -> impl Fn(I) -> IResult<I, Option<O>>
where
    F: Fn(I) -> IResult<I, O>,
{
    move |i: I| {
        let (i, x) = f(i)?;
        Ok((i, Some(x)))
    }
}

pub fn none<'a, I, O, F>(_f: F) -> impl Fn(I) -> IResult<I, Option<O>>
where
    F: Fn(I) -> IResult<I, O>,
{
    move |i: I| Ok((i, None))
}

pub fn alt_left<'a, I, O, F, G>(f: F, _g: G) -> impl Fn(I) -> IResult<I, O>
where
    F: Fn(I) -> IResult<I, O>,
    G: Fn(I) -> IResult<I, O>,
{
    move |i: I| {
        let (i, x) = f(i)?;
        Ok((i, x))
    }
}

pub fn alt_right<'a, I, O, F, G>(_f: F, g: G) -> impl Fn(I) -> IResult<I, O>
where
    F: Fn(I) -> IResult<I, O>,
    G: Fn(I) -> IResult<I, O>,
{
    move |i: I| {
        let (i, x) = g(i)?;
        Ok((i, x))
    }
}
